### PR TITLE
Implement Logbox module

### PR DESCRIPTION
**Pull request by acoates-ms**
Logbox is a replacement of the YellowBox UI, used to display warnings.  It is currently available by calling 
```js
require('react-native').unstable_enableLogBox();
```

but it will be enabled by default in 0.63.

LogBox works by creating an additional ReactRootView, so it will provide some good test coverage of multi root view scenarios.  I also fixed an issue with layout when there are multiple rootviews, where we would attempt to layout a bunch of elements before they had layout run on them.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5175)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/Modules/LogBoxModule.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  REACT_INIT(Initialize)
+  void Initialize(React::ReactContext const &reactContext) noexcept;
+
+  REACT_METHOD(show) void show() noexcept;
+  REACT_METHOD(hide) void hide() noexcept;
+
+ private:
+  void showOnUIThread() noexcept;
+  void hideOnUIThread() noexcept;
+  void onPopupClosed() noexcept;
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/Modules/LogBoxModule.h:23]: can you please keep the casing of C++ method consistent (CamelCase)

File: vnext/Microsoft.ReactNative/ReactNativeHost.cpp [hunk 2]
```diff
@@ -23,7 +23,9 @@ using namespace xaml::Controls;
 
 namespace winrt::Microsoft::ReactNative::implementation {
 
-ReactNativeHost::ReactNativeHost() noexcept : m_reactHost{Mso::React::MakeReactHost()} {
+ReactNativeHost::ReactNativeHost() noexcept : ReactNativeHost(*Mso::React::MakeReactHost()) {}
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/ReactNativeHost.cpp:26]: std::move? since this is a temporary that gets destroyed immediately

**Comment by asklar** [comment]
> Logbox is a replacement of the YellowBox UI, used to display warnings. It is currently available by calling
> 
> ```js
> require('react-native').unstable_enableLogBox();
> ```
> 
> but it will be enabled by default in 0.63.
> 
> LogBox works by creating an additional ReactRootView, so it will provide some good test coverage of multi root view scenarios. I also fixed an issue with layout when there are multiple rootviews, where we would attempt to layout a bunch of elements before they had layout run on them.
> 
> ###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5175)

we call the yellowbox API when we detect an unimplemented property in Debug. Do we need any changes in that codepath (which just calls js) to plug into this scenario?

**Comment by TheSavior** [comment]
You aren’t going to include a cool screenshot in your summary? ;-)

**Comment by acoates-ms** [comment]
Cool screenshot:

**Comment by acoates-ms** [comment]
> we call the yellowbox API when we detect an unimplemented property in Debug. Do we need any changes in that codepath (which just calls js) to plug into this scenario?

That API should still work.

File: vnext/Microsoft.ReactNative/Modules/LogBoxModule.cpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  if (!instance)
+    return;
+  auto weakHost = instance->GetHost();
+  auto host = weakHost.GetStrongPtr();
+  if (!host)
+    return;
+
+  m_logBoxContent = React::ReactRootView();
+  m_logBoxContent.ComponentName(L"LogBox");
+  m_logBoxContent.ReactNativeHost(winrt::make<winrt::Microsoft::ReactNative::implementation::ReactNativeHost>(*host));
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/Modules/LogBoxModule.cpp:49]: >make [](start = 41, length = 4)

We must not create a new ReactNativeHost. Instead, we should expose it from the ReactContext and keep ReactNativeHost as a weak ptr inside of the ReactHost.

File: vnext/Microsoft.ReactNative/ReactNativeHost.cpp [hunk 4]
```diff
@@ -52,6 +52,24 @@ IAsyncAction ReactNativeHost::LoadInstance() noexcept {
...
+ReactNativeHostProperty() noexcept {
+  static ::Microsoft::ReactNative::ReactPropertyId<
+      ::Microsoft::ReactNative::ReactNonAbiValue<winrt::weak_ref<ReactNativeHost>>>
+      propId{L"ReactNative.ReactNativeHost", L"ReactNativeHost"};
+  return propId;
+}
+
+/*static*/ winrt::Microsoft::ReactNative::ReactNativeHost ReactNativeHost::GetReactNativeHost(
+    ReactPropertyBag const &properties) noexcept {
+  if (auto wkHost = properties.Get(ReactNativeHostProperty()).Value()) {
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/ReactNativeHost.cpp:65]: >.Value() [](start = 61, length = 8)

We should account for a missing property value.
The simplest way is to add another "if (auto ..."

File: vnext/Microsoft.ReactNative/RedBox.cpp [hunk 5]
```diff
@@ -309,13 +309,14 @@ struct RedBox : public std::enable_shared_from_this<RedBox> {
...
     for (const auto frame : m_errorInfo.Callstack) {
       const winrt::hstring xamlFrameString =
-          L"<StackPanel Margin='0,5,0,5'"
-          L"  xmlns='http://schemas.microsoft.com/winfx/2006/xaml/presentation'"
-          L"  xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'"
-          L"  >"
-          L"  <TextBlock x:Name='MethodText' FontSize='14' Foreground='White' TextWrapping='Wrap' FontFamily='Consolas'/>"
-          L"  <TextBlock x:Name='FrameText' FontSize='14' Foreground='Gray' TextWrapping='Wrap' FontFamily='Consolas'/>"
-          L"</StackPanel>";
+          LR"(
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/RedBox.cpp:312]: >LR"( [](start = 10, length = 4)

Consider changing it to LR"XAML( ... )XAML" to increase readability and uniqueness of the delimiters.

File: vnext/Microsoft.ReactNative/RedBox.cpp [hunk 6]
```diff
@@ -60,31 +60,32 @@ struct RedBox : public std::enable_shared_from_this<RedBox> {
...
+            <ColumnDefinition Width='*'/>
+            <ColumnDefinition Width='*'/>
+          </Grid.ColumnDefinitions>
+          <Grid.RowDefinitions>
+            <RowDefinition Height='*'/>
+            <RowDefinition Height='Auto'/>
+          </Grid.RowDefinitions>
+          <ScrollViewer Grid.Row='0' Grid.ColumnSpan='2' HorizontalAlignment='Stretch'>
+            <StackPanel HorizontalAlignment='Stretch'>
+              <StackPanel Background='#FFF35369' HorizontalAlignment='Stretch' Padding='15,35,15,15' x:Name='StackPanelUpper'>
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/RedBox.cpp:78]: is this right? this color is a mustard yellow, but the redbox is supposed to be for errors...and it has red in the name :)

Comment (TheSavior) [reply vnext/Microsoft.ReactNative/RedBox.cpp:78]: The new Logbox replaced Redbox and Yellowbox. I think we still have a native redbox for errors that occur when loading JS that can't render the Logbox. Is that what this is? If so, I wouldn't expect a yellow color in native code (although my understanding of Logbox / redbox is surely incomplete).

Comment (asklar) [reply vnext/Microsoft.ReactNative/RedBox.cpp:78]: I want to make sure that errors coming from metro etc, are shown distinctly as errors. Here's what we have today for one type of error from metro.
image

Comment (acoates-ms) [reply vnext/Microsoft.ReactNative/RedBox.cpp]: Its not a yellow, its a pale red.  I was updating the native redbox colors to match the ones used by logbox errors.  LogBox errors use a lighter red.  But taking another look at core it looks like core native redbox doesn't use the same color as logbox errors.  But they did update the color of the native redbox.  -- So I'll change to use that color. (#D01926)

https://github.com/facebook/react-native/blob/40b36a040f06aae03be1a09f14e48eb0c652de96/ReactAndroid/src/main/res/devsupport/layout/redbox_item_title.xml#L10

**Pull request merged**